R/extract_drug_dispenses.R

```diff
@@ -77,41 +79,61 @@ extract_drug_dispenses <- function(
     dis_dtd_lag_months = 7,
     patients_ids = NULL,
     output_table_name = NULL,
+    overwrite = FALSE,
     conn = NULL) {
-  if (is.null(start_date) || is.null(end_date)) {
-    stop("Both start_date and end_date must be provided.")
-  }
-
-  if (!inherits(start_date, "Date") || !inherits(end_date, "Date")) {
-    stop("start_date and end_date must be Date objects.")
-  }
-
-  if (start_date > end_date) {
-    stop("start_date must be earlier than or equal to end_date.")
-  }
+  stopifnot(
+    !is.null(start_date),
+    !is.null(end_date),
+    inherits(start_date, "Date"),
+    inherits(end_date, "Date"),
+    start_date <= end_date
+  )
 
   connection_opened <- FALSE
   if (is.null(conn)) {
-    conn <- initialize_connection()
+    conn <- connect_oracle()
     connection_opened <- TRUE
   }
 
   if (!is.null(output_table_name)) {
+    output_table_name_is_temp <- FALSE
     if (!is.character(output_table_name)) {
       stop("output_table_name must be a character string")
     }
-    table_name <- output_table_name
-    if (DBI::dbExistsTable(conn, table_name)) {
-      warning(paste("Table", table_name, "already exists. It will be overwritten."))
+    if (DBI::dbExistsTable(conn, output_table_name)) {
+      if (overwrite) {
+        warning(
+          glue::glue(
+            "Table {output_table_name} already exists.
+            It will be overwritten."
+          )
+        )
+      } else {
+        stop(
+          glue::glue(
+            "Table {output_table_name} already exists.
+            Set overwrite = TRUE to overwrite it."
+          )
+        )
+      }
     }
   } else {
-    table_name <- paste0("TMP_DISP_", format(Sys.time(), "%Y%m%d_%H%M%S"))
+    output_table_name_is_temp <- TRUE
+    timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
+    output_table_name <- glue::glue("TMP_DISP_{timestamp}")
   }
   try(
-    DBI::dbRemoveTable(conn, table_name),
+    DBI::dbRemoveTable(conn, output_table_name),
     silent = TRUE
```

Comment:
Je pense qu'il ne faut en aucun cas supprimer une table qu'on n'a pas crée nous même.

Je pense que tu prends assez de précautions avec le `timestamp` pour qu'il soit extrêmement peu probable que la table existe. Si la table existe, `dbWriteTable()` retournera une erreur et l'utilisateur devra prendre lui-même la décision de supprimer la table s'il le souhaite. Au pire, il ne comprend pas l'erreur, il relance la fonction, le `timestamp` change et le problème est résolu tout seul.

Par contre il faut bien faire le ménage de toutes les tables "temporaires" que l'on crée à la fin de fonction : `dbRemoveTable(conn, output_table_name)` si l'utilisateur veut juste un data frame.

Reply:
Umh ce commentaire n'était peu être pas pertinent pour `output_table_name` en effet. Mais pourquoi tu ne gères pas `overwrite` dans `dbWriteTable()` directement ? (i.e. `dbWriteTable(*, overwrite = overwrite)`) (C'est peut-être une question bête, j'ai pas le temps de m'y pencher maintenant.)

Reply:
C'est ce que je compte faire si on le garde. Je remets l'option `overwrite` que je venais d'enlever du coup

Reply:
Sorry ! 😳

Reply:
Finalement le overwrite est géré au début du code : 
````

  if (!is.null(output_table_name)) {
    output_table_name_is_temp <- FALSE
    stopifnot(
      is.character(output_table_name),
      !DBI::dbExistsTable(conn, output_table_name) || (DBI::dbExistsTable(conn, output_table_name) && overwrite)
    )
    if (DBI::dbExistsTable(conn, output_table_name) && overwrite) {
      warning(
        glue::glue(
          "Table {output_table_name} already exists and will be overwritten."
        )
      )
      DBI::dbRemoveTable(conn, output_table_name)
    }
  }
````
On ne pouvait pas le mettre dans la boucle tel quel avec une simple condition `DBI::dbExistsTable(conn, output_table_name)` car sinon `dbWriteTable(*, overwrite = overwrite)` réécrirait par dessus l'itération précédente de la boucle à chaque itération...